new windows module win_audit_policy_system

nwsparks opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

- New module for managing system wide audit policy on Windows. 
- Includes backup/restore capability
- Does not include user level audit management. It was in there initially but due to the complications it was causing I decided it would be best if it was moved to it's own module. I'll follow up at some point with another module for it. This is the reason for the module naming.
- No explicit modify in integration testing because "remove" is checking both things. Policies are never actually removed, what changes is just the type of events that are being recorded.
- There wasn't a short way of retrieving existing rules for the integration testing so the testing module was required. It's just returning a dictionary mapping the policy in question to the state of the logging for it (success, failure, both, no auditing). Assertions are then done against it.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request


##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
win_audit_policy_system
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
[root@ip-10-5-10-122 ansible]# ansible --version
ansible 2.5.0 (win_audit_policy e8b881a977) last updated 2017/09/28 13:19:13 (GMT +000)
  config file = None
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /opt/ansible/lib/ansible
  executable location = /opt/ansible/bin/ansible
  python version = 2.7.12 (default, Sep  1 2016, 22:14:00) [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)]

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

nwsparks committed:
new windows module win_audit_policy_system

jborean93 commented on lib/ansible/modules/windows/win_audit_policy_system.ps1:
```diff
@@ -0,0 +1,184 @@
+$supplied_params = $params.GetEnumerator() | Where-Object {$_.key -notlike "_*"}
+
+#do nothing and continue
+If ($backup_path -and $restore_path)
```
There are 2 no-op branches in this if statement, why don't you just check the 2 if statement's that do something and just check the categories/sub categories if they are set.

jborean93 commented on lib/ansible/modules/windows/win_audit_policy_system.ps1:
```diff
@@ -0,0 +1,184 @@
+
+    If ($backup_only = $true)
+    {
+        $results.changed = $true
```
You should be setting `$result.changed = $true` straight after you run the auditpol executable.

jborean93 reviewed (commented):
I still need to test this manually but I have a few thoughts on this

* I don't think the restore/backup options should be a function of this module. 

It doesn't really control the state of the audit policy but rather performs an action. It also is just a wrapper around calling the executable to run a particular task which can easily be done with `win_command`. Personally I think they should be removed to make the module more streamlined, and more purpose built to set the audit policy. Happy to be proven otherwise but I'll let other people comment on this.

* I am not a fan of calling `auditpol.exe` to perform these tasks, in the past calling executables has given us some grief especially when dealing with other languages.

Having a quick glance online there is no cmdlet/wmi or com object that we can easily call but we can use the Win32 API's through P/Invoke if necessary https://msdn.microsoft.com/en-us/library/aa375742(VS.85).aspx. P/Invoke does bring some more verbosity/complexity compared to cmdlets/wmi or com objects but it does make calling these tasks a lot simpler than doing string parsing on executable outputs.

I don't know how much experience you've had with calling the native Win32 API's from Powershell but that might be a better option if you want to take a go at it. There are a few modules currently where we do this, look at win_user_right if you want to see how it is done. Having a look at what is being performed it seems like there are a few calls that may need to be run but it might make managing policies easier in the future.

You may even want to consider using the `Run-Command` cmdlet in `Ansible.ModuleUtils.CommandUtil` to call the executable if you want to continue with using auditpol.exe. `Run-Command` can take in a string to execute and it returns a hashtable that contains the rc, stdout and stderr as separate objects. This makes it easier to validate if the command has run than when it is just called straight from Powershell.

Are you planning on creating separate modules for managing audit policies on the user level?

nwsparks commented:
Thanks @jborean93 

I don't have issue with removing backup/restore. I do think there is some value in it being there in that it reminds people of the ability to perform those actions, but agree removing it makes things more streamlined. 

I did take a stab at one point with the win32 api, but wasn't able to get it to work...the effort I put in was pretty minimal though so I can try and revisit it.

I had planned to follow up with a separate module for the user specific settings. initially I had included it in this module, but it led to having a ton of mutually exclusive parameters due to how auditpol functions which made the module feel super clunky. There was a decent bit of weird behavior around this area of auditpol as well. If I'm able to get win32 api's working I might look at re-including it since the problems were mostly just stemming from auditpol.

I guess for now I'll try and look at the win32 api stuff and either try and get that implemented or if it still proves too troublesome I can follow up with a revision that uses run-command as you suggested.

jborean93 commented:
If you can't get the Win32 API working I don't want to hold this up, just let me know how you go and if you still want to continue the current track.

nwsparks commented:
Ok. I don't really have any experience with .net or the api so it's been a bit of a struggle. I have portions of it done, but don't think I'd have a finished product any time soon.

I'll see if I can rework to use run-command in the next few days and recommit. I'll also be removing the backup/restore functions and fix the items you mentioned above.

nwsparks committed:
removed the backup/restore functionality

adjusted to use run-command rather than running the command directly

adjusted testing appropriately for the above changes

nwsparks committed:
fixed issue with variable naming in testing
removed .psm1 from requires
fixed copyright

nwsparks commented:
Ok this should be good to go. LMK if any issues. ready_for_review

nwsparks commented:
This one's been sitting for a bit. Did you guys want to see anymore changes?

dagwieers commented on lib/ansible/modules/windows/win_audit_policy_system.ps1:
```diff
@@ -0,0 +1,129 @@
+#!powershell
+# Copyright (c) 2017 Noah Sparks <nsparks@outlook.com>
```
Just being pedantic here, not a showstopper. But the syntax for copyright statement was changed to:
```powershell
# Copyright: (c) 2017, Noah Sparks <nsparks@outlook.com>
# Copyright: (c) 2017, Ansible Project
```

nwsparks replied:
Np. I'll change since it looks like there's some new validation it failed for other items.

dagwieers commented on lib/ansible/modules/windows/win_audit_policy_system.ps1:
```diff
@@ -0,0 +1,129 @@
+}
+
+################################
+### Deal with parameter sets ###
```
I personally don't like that style for comments. And it may trigger issues when we start doing style validation in the future ;-)

nwsparks replied:
thanks for checking it. You'd prefer just simple like this? 

`# comment here`

I like the fatter blocks to split sections up, but I don't mind changing it.

dagwieers replied:
I'll leave it up to others to decided, as it is a personal opinion. (That's why I didn't ask for a change in this case).

nwsparks replied:
Ok gotcha

dagwieers reviewed (approved):
LGTM

jamesla commented:
Thank you for writing this hopefully it gets merged soon :-)

jborean93 commented on lib/ansible/modules/windows/win_audit_policy_system.py:
```diff
@@ -0,0 +1,74 @@
+'''
+
+RETURN = '''
+backup_taken:
```
Is this required anymore now that it is removed from the module.

nwsparks replied:
good catch

jborean93 commented on lib/ansible/modules/windows/win_audit_policy_system.py:
```diff
@@ -0,0 +1,74 @@
+  audit_type:
+    description:
+      - The type of event you would like to audit for.
+    choices: [ 'success', 'failure', 'success and failure', 'none' ]
```
The more I think about this the more I think this should be a list type where you can specify some the options `success` and `failure`. You can use a combination of both (`success and failure`) or just an empty list (`none`) for the other options. We should also decide on a default, potentially `none` is the default if it isn't set.

nwsparks replied:
I think I did it this way to make the values match what auditpol outputs so I wouldn't have to do string manipulation. I can take a look though and see about changing it to a list.

I wonder if making it required instead of a default would be better here to force a selection. Otherwise, for a default, I think 'success' would make the most sense. I think people using this module would typically looking to be enable auditing rather than disable. Either way though you're right, it does need one of these. I'll give some thought about it...I guess there isn't much difference between the two really.

jborean93 replied:
Sounds good to me

jborean93 reviewed (commented):
Only have 2 points so far but the rest looks great.

nwsparks committed:
Updated audit_type to list and added appropriate error handling
Updated testing accordingly
Fixed up documentation

nwsparks commented:
@jborean93 ok suggested changes have been made. docs updated, audit_type changed to list. I ended up doing failifempty for the audit_types. I think due to the nature of this, even if minor, making people take a little extra consideration for what they are adjusting would be good. If you really prefer the default though I can change it to -default success

jborean93 commented:
LGTM, are you able to update the CHANGELOG once it is merged in.

rebuild_merge

nwsparks commented:
thanks. will update changelog now.
